pkg/report/coverageprofile.go

```diff
@@ -0,0 +1,306 @@
+package report
+
+import (
+	"fmt"
+	"regexp"
+	"sort"
+	"strings"
+
+	"github.com/Azure/gocover/pkg/gittool"
+	"golang.org/x/tools/cover"
+)
+
+// DiffCoverage expose the diff coverage statistics
+type DiffCoverage interface {
+	GenerateDiffCoverage() (*Statistics, error)
+}
+
+func NewDiffCoverage(
+	profiles []*cover.Profile,
+	changes []*gittool.Change,
+	excludes []string,
+	comparedBranch string,
+) DiffCoverage {
+	return &diffCoverage{
+		comparedBranch: comparedBranch,
+		excludes:       excludes,
+		profiles:       profiles,
+		changes:        changes,
+	}
+}
+
+var _ DiffCoverage = (*diffCoverage)(nil)
+
+// diffCoverage implements the DiffCoverage interface
+// and generate the diff coverage statistics
+type diffCoverage struct {
+	comparedBranch string            // git diff base branch
+	excludes       []string          // excludes files
+	profiles       []*cover.Profile  // go unit test coverage profiles
+	changes        []*gittool.Change // diff change between compared branch and HEAD commit
+}
+
+func (diff *diffCoverage) GenerateDiffCoverage() (*Statistics, error) {
+	if err := diff.ignore(); err != nil {
+		return nil, err
+	}
+	diff.filter()
+	return diff.percentCovered(), nil
+}
+
+// ignore files that not accountting for diff coverage
+// support standard regular expression
+func (diff *diffCoverage) ignore() error {
+	var filterProfiles []*cover.Profile
+
+	for _, p := range diff.profiles {
+		find := false
+		for _, ignorePattern := range diff.excludes {
+			reg, err := regexp.Compile(ignorePattern)
+			if err != nil {
+				return fmt.Errorf("compile pattern %s: %w", ignorePattern, err)
+			}
+			if reg.MatchString(p.FileName) {
+				find = true
+				break
+			}
+		}
+		if !find {
+			filterProfiles = append(filterProfiles, p)
+		}
+	}
+
+	diff.profiles = filterProfiles
+	return nil
+}
+
+// filter files that no change in current HEAD commit
+func (diff *diffCoverage) filter() {
+	var filterProfiles []*cover.Profile
+	for _, p := range diff.profiles {
+		for _, c := range diff.changes {
+			if strings.HasSuffix(p.FileName, c.FileName) {
```

Comment:
question: why using `HasSuffix` here? Will pattern `ab` includes `abc` ? But actually they are different files, right?

Reply:
In the cover profile generated by `go test`, the tested go filepath includes module name, and the `git diff` command generates the filepath relative to the project, it means that a go file in cover profile is like this `{module}\pkg\a.go`, while `gi diff` is `pkg\a.go`. This use the `HasSuffix` at this context.

Reply:
what will happen for case

```
{module}\pkg\a   # a file with name a
{module}\pkg\a.go
```

Will the first item be the suffix of the second item?

Reply:
Both `strings.HasSuffix("{module}/pkg/a", "{module}/pkg/a.go")` and  `strings.HasSuffix("{module}/pkg/a.go", "{module}/pkg/a")` are false.

And `strings.HasSuffix("{module}/pkg/a.go", "pkg/a.go")` is true.

```
// HasSuffix tests whether the string s ends with suffix.
func HasSuffix(s, suffix string) bool {
```

Reply:
oh I see, I misread the suffix to prefix. How about this case:

```
a.go
go
```

The general idea is comparing the whole string might have chance to see partial matching. We probably want to compare the file path components one by one to avoid such case. This is more or less a tree searching problem instead of string matching.

Reply:
Also, to clarify the intention, I would suggest we rename the file path checking to a function `isSubFolderTo(parentDir, filepath)`. This could help the reviewer and the developer who changes the code in the future.